Assets/AY_Scripts/TankController.cs

```diff
@@ -2,33 +2,84 @@
 using System.Collections.Generic;
 using UnityEngine;
 
-public class TankController : MonoBehaviour
+public class TankController : MonoSingletonGeneric<TankController>
 {
-    public float speed = 10;
-
+    public GameObject tankBody;
+    public ShellController shell;
+    
+    /*---Private---*/
     private Joystick joystick;
-    private Transform tankParent;
+
+    /*---Tank Properties---*/
+    private TextMesh floatingName;
+    private string tankName;
+    private float speed;
+    private int damage;
+    private int health;
+
+    /*---Movement---*/
     private float horizontalMove;
     private float verticalMove;
     private Vector3 lookDirection;
+    private Rigidbody rb;
+    private Vector3 movement;
     
-    void Awake()
+    public void Initialize(TankTypesScriptable t, Joystick jt)
+    {
+        tankName = t.tankName;
+        speed = t.speed;
+        damage = t.damage;
+        health = t.health;
+        joystick = jt;
+        Debug.Log("Player : " + tankName + " Speed : " + speed);
+    }
+    protected override void Awake()
     {
-        joystick = FindObjectOfType<FixedJoystick>();
-        tankParent = transform.parent;
+        base.Awake();
+        rb = GetComponent<Rigidbody>();
     }
+
+    private void Start()
+    {
+        floatingName = GetComponentInChildren<TextMesh>();
+        floatingName.text = tankName;
+    }
+
     void Update()
+    {
+        Movement();
+        NameLookAtCamera();
+        TankService.Instance.shellPosition = tankBody;
+    }
+
+    private void Movement()
     {
         horizontalMove = joystick.Horizontal * speed;
         verticalMove = joystick.Vertical * speed;
-        Vector3 pos = tankParent.position;
-        pos.x += horizontalMove * Time.deltaTime;
-        pos.z += verticalMove * Time.deltaTime;
-        tankParent.position = pos;
+        movement = new Vector3(horizontalMove, 0, verticalMove);
+        rb.MovePosition(transform.position + movement * Time.deltaTime);
+        Turn();
+    }
+
+    private void Turn()
+    {
         if (joystick.Direction.x != 0 && joystick.Direction.y != 0)
         {
             lookDirection = new Vector3(joystick.Direction.x, 0, joystick.Direction.y);
         }
-        transform.rotation = Quaternion.LookRotation(lookDirection, Vector3.up);
+        tankBody.transform.rotation = Quaternion.LookRotation(lookDirection, Vector3.up);
+    }
+
+    private void NameLookAtCamera()
+    {
+        floatingName.transform.LookAt(Camera.main.transform.position);
+        floatingName.transform.Rotate(0, 180, 0);
+    }
+
+    public void Shoot()
+    {
+        ShellController shellObject = Instantiate(shell, TankService.Instance.shellPosition.transform.position, TankService.Instance.shellPosition.transform.rotation);
+        Rigidbody shellRB = shellObject.GetComponent<Rigidbody>();
```

Comment:
Make a function called "MOVE" in shellController and pass in values from the tank on Shoot